std/sumtype.d

```diff
@@ -1986,7 +1996,26 @@ private template matchImpl(Flag!"exhaustive" exhaustive, handlers...)
                 case caseId:
                     static if (matches[caseId] != noMatch)
                     {
-                        return mixin(handlerName!(matches[caseId]), "(", handlerArgs!caseId, ")");
+                        alias matchedHandler = handlerName!(matches[caseId]);
+                        static immutable matchedCall = matchedHandler ~ "(" ~ handlerArgs!caseId ~ ")";
+
+                        import std.traits : ParameterStorageClassTuple, ParameterStorageClass, isCallable, isUnsafe;
+                        alias STC = ParameterStorageClass;
+
+                        static if (
+                            isCallable!(mixin(matchedHandler)) &&
+                            (ParameterStorageClassTuple!(mixin(matchedHandler))[0] == STC.ref_ ||
+                                isUnsafe!(mixin(matchedHandler)))
+                            )
+                        {
+                            return mixin(matchedCall);
+                        }
+                        else
+                        {
+                            return () @trusted inout {
+                                return mixin(matchedCall);
+                            }();
+                        }
```

Comment:
`isCallable` returns `false` for templates, so this will allow `@safe` code to call a `@system` handler as long as that handler is a template.

Reply:
We can detect if it is a template. Currently, traits to get the storage class doesn't support templates. Would be much more straightforward to do this if `__traits(getParameterStorageClasses)` supports `CallExp`. I'm currently investigating this on the compiler side, but even with forward references for types, it is possible to fetch the parameters storage class.

Reply:
https://github.com/dlang/dmd/pull/12712

Reply:
This has been fixed in: dlang/dmd#12712